agent/connect/ca/aws_pca.go

```diff
@@ -0,0 +1,481 @@
+package ca
+
+// Assumptions that need to be changed:
+// 1. ECDSA is accepted everywhere. AWS only supports RSA.
+// 2. CA's are able to cross-sign. Hosted CA's don't seem to allow this.
+// 3. Certs don't require a Subject. AWS appears to require a CN.
+// 4. Validity periods can be as short as 1h. AWS minimum is 1d.
+
+//Current issues and problems:
+// 1. Vault provider does not work if configured at bootstrap time. You have to boostrap with Consul and then switch to Vault.
+//    Error: no handler for route 'consul-intermediate/sign/leaf-cert'
+// 2. With any provider besides Consul, the leaf certs are constantly refreshed because the cache is storing the
+//    subject key of the root cert, but Vault and AWS are signing with an intermediate cert.
+
+import (
+	"crypto/x509"
+	"fmt"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
+	"regexp"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/acmpca"
+
+	"github.com/hashicorp/consul/agent/connect"
+)
+
+const (
+	RootTemplateARN         = "arn:aws:acm-pca:::template/RootCACertificate/V1"
+	IntermediateTemplateARN = "arn:aws:acm-pca:::template/SubordinateCACertificate_PathLen0/V1"
+	LeafTemplateARN         = "arn:aws:acm-pca:::template/EndEntityCertificate/V1"
+)
+
+const (
+	RootValidity         = 5 * 365 * 24 * time.Hour
+	IntermediateValidity = 1 * 365 * 24 * time.Hour
+)
+
+type AmazonPCA struct {
+	arn              string
+	pcaType          string
+	keyAlgorithm     string
+	signingAlgorithm string
+	sleepTime        time.Duration
+
+	certPEM  string
+	chainPEM string
+	//signingKey  string
+	client *acmpca.ACMPCA
+	logger hclog.Logger
+}
+
+func createPCALogger(pcaType string) hclog.Logger {
+	return hclog.New(&hclog.LoggerOptions{
+		Name:  "pca-" + pcaType,
+		Level: hclog.Trace,
+	})
```

Comment:
See other PR - we can't use hclog yet and need to pass in a logger from the server to make sure format and output destination are preserved. I proposed a solution on the other PR.

Reply:
Fixed by 3a530b18f33a99f79d4fa1fe31969b4f4973fea6.